imap/caldav_alarm.h

```diff
@@ -92,9 +98,48 @@ int caldav_alarm_delete_mailbox(const char *mboxname);
 /* delete all alarms for a user */
 int caldav_alarm_delete_user(const char *userid);
 
-/* distribute alarms with triggers in the next minute */
+/* process alarms with triggers before a given time */
 int caldav_alarm_process(time_t runtime, time_t *next, int dryrun);
 
+/* list alarms (via callbacks) before a given time */
+typedef void (*list_calendar_proc)(const mbname_t *mbname,
+                                   uint32_t imap_uid,
+                                   time_t nextcheck,
+                                   uint32_t num_rcpts,
+                                   uint32_t num_retries,
+                                   time_t last_run,
+                                   const char *last_err,
+                                   void *rock);
+typedef void (*list_snooze_proc)(const mbname_t *mbname,
+                                 time_t nextcheck,
+                                 uint32_t num_retries,
+                                 time_t last_run,
+                                 const char *last_err,
+                                 json_t *snoozed,
+                                 void *rock);
+typedef void (*list_send_proc)(const mbname_t *mbname,
+                               time_t nextcheck,
+                               uint32_t num_retries,
+                               time_t last_run,
+                               const char *last_err,
+                               json_t *submission,
+                               void *rock);
+typedef void (*list_unscheduled_proc)(const mbname_t *mbname,
+                                      uint32_t imap_uid,
+                                      time_t nextcheck,
+                                      uint32_t num_rcpts,
+                                      uint32_t num_retries,
+                                      time_t last_run,
+                                      const char *last_err,
+                                      void *rock);
+int caldav_alarm_list(time_t runtime,
```

Comment:
Yeah... I consider this `foo_list()` stuff a temporary API because it's specifically only for listing them.  Theoretically at some point we refactor this all one level deeper, where "list" or "process" are themselves callbacks or options or something, and then the outer API becomes a foreach API.

But the refactor to get there is too big and messy to do on a whim, so for now there's a specific listing API, and it's called "list" because that's what it does.

Reply:
I still find using "list" rather than the more common "foreach" confusing, but I'll manage with either name.